### PR TITLE
Hook implementation for event object display names

### LINKED ISSUE #132: Implement hooks to render objects
With the new changes in notifications, the object table will no longer have the `host`, `service` and will support all types of objects. See https://github.com/Icinga/icinga-notifications/issues/85

This makes it difficult to identify the object, whether it is a host, a service or a certificate. The search bar cannot explicitly search for hosts and services. Rendering lists (event, incident) no longer works as before.

There should be some hooks that can do the job based on the source.

### PR DESCRIPTION
**Pull request by raviks789**
resolves #132 

Blocked by: #155, #166

requires: https://github.com/Icinga/ipl-web/pull/217

### CODE REVIEW

**Comment by nilmerg** [comment]
Please rebase, fix conflicts and mark as ready for review.

**Comment by sukhwinder33445** [comment]
Please rebase and resolve the conflicts.

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]

File: library/Notifications/Widget/ItemList/EventList.php [hunk 1]
```diff
@@ -27,10 +38,36 @@ public function __construct(ResultSet $data)
...
+    /**
+     * Get the rendered object display name for the given object ID
+     *
+     * @param string $objectID
+     *
+     * @return ?ValidHtml
+     */
+    public function getRenderedObjectDisplayName(string $objectID): ?ValidHtml
+    {
+        return $this->renderedObjectDisplayNames[$objectID] ?? null;
```
[the remark this answers is not in the document]
Comment (raviks789) [reply library/Notifications/Widget/ItemList/EventList.php:71]: This is a dead code.

File: library/Notifications/Hook/EventsObjectsInfoHook.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+     * @return ?ValidHtml
+     */
+    final public static function getObjectListItemWidget(string $source, array $objectIdTag): ?ValidHtml
+    {
+        $objectListItemWidget = null;
+        foreach (Hook::all('Notifications\\EventsObjectsInfo') as $hook) {
+            /** @var self $hook */
+            try {
+                if ($source === $hook->getSource()) {
+                    $objectListItemWidget = $hook->getHtmlForObject($objectIdTag);
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Notifications/Hook/EventsObjectsInfoHook.php]: You have still not returned in the loop. Should the first `$objectListItemWidget` found be overwritten if there is more than one hook?

[the remark this answers is not in the document]
Comment (raviks789) [reply library/Notifications/Hook/EventsObjectsInfoHook.php]: No, every object as a single source so this will not be replaced and that is also the reason for checking whether source's hook matches with object's hook. But I guess it is still better to return here.

[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Notifications/Hook/EventsObjectsInfoHook.php]: Yes, because further iterations are useless.

File: library/Notifications/Hook/EventsObjectsInfoHook.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+     * @return array<string, ValidHtml>
+     */
+    abstract public function getObjectNameWidgets(array $objectIdTags): array;
+
+    /**
+     * Get the source of the objects
+     *
+     * @return string
+     */
+    abstract public function getSource(): string;
```
Comment (sukhwinder33445) [inline library/Notifications/Hook/EventsObjectsInfoHook.php:35]: It returns the source type, name the method accordingly, please.

File: library/Notifications/Hook/EventsObjectsInfoHook.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+{
+    /** @var array<string, ValidHtml>  */
+    protected static $objectNameWidgets = [];
+
+    /**
+     * Get the object name widgets for the objects using the object ID tags
+     *
+     * @param array<string, array<string, string>> $objectIdTags
+     *
+     * @return array<string, ValidHtml>
```
Comment (sukhwinder33445) [inline library/Notifications/Hook/EventsObjectsInfoHook.php:26]: This type hint seems to be incorrect.
It should be `@return array<string, array<string, ValidHtml>` // Outer array key must be source type, inner array key must be object id. Please document this.

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+     * @return ValidHtml
+     */
+    abstract public function createObjectLink(array $objectIdTag): ValidHtml;
+
+    /**
+     * Register HTMLs to be rendered for the object names using the cached object ID tags
+     *
+     * @return void
+     */
+    final public static function registerHtmlForObjectNames(): void
```
Comment (sukhwinder33445) [inline library/Notifications/Hook/ObjectsRendererHook.php:68]: Please wrap the function body in a `new private static method` and call it here.  his new method can then also be called in `getObjectName` if this registration does not take place. For example, if the object is not a list item.

Function name : e.g. `prepare(Array)`

[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Notifications/Hook/ObjectsRendererHook.php:68]: Please change the name of the property accordingly.

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+     * @param Objects $obj
+     *
+     * @return ValidHtml
+     */
+    final public static function getObjectName(Objects $obj): ValidHtml
+    {
+        $objId = $obj->id;
+        if (isset(self::$registeredObjectNameHtmls[$objId])) {
+            return self::$registeredObjectNameHtmls[$objId];
+        }
```
Comment (sukhwinder33445) [inline library/Notifications/Hook/ObjectsRendererHook.php:98]: And here we can call the new method.
```suggestion
        if (! isset(self::$registeredObjectNameHtmls[$objId])) {
            $data[$obj->source->type][$objId] = $obj->id_tags;
            self::prepare($data);
        }
        
        if (isset(self::$registeredObjectNameHtmls[$objId])) {
            return self::$registeredObjectNameHtmls[$objId];
        }
```

File: library/Notifications/Widget/ItemList/EventListItem.php [hunk 7]
```diff
@@ -105,14 +107,15 @@ protected function assembleTitle(BaseHtmlElement $title): void
 
         /** @var Objects $obj */
         $obj = $this->item->object;
-        $name = $obj->getName();
+        $name = (new DeferredText(function () use ($obj) {
```
Comment (sukhwinder33445) [inline library/Notifications/Widget/ItemList/EventListItem.php:110]: `DefferedText` is not required here, as the name has already been rendered. Please undo the change.

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]

File: library/Notifications/Hook/ObjectsRendererHook.php (hunk 5 above)
Comment (raviks789) [reply library/Notifications/Hook/ObjectsRendererHook.php:68]: If an HTML for the object name is not already loaded, I don't think doing this has any effect. If you are suggesting this because of Event/Incident detail view shown in a single column. It would still be unnecessary as the object is registered and the HTMLs for the object is already prepared. If not then it should be prepared from the object ID tags.

File: library/Notifications/Hook/ObjectsRendererHook.php (hunk 6 above)
Comment (raviks789) [reply library/Notifications/Hook/ObjectsRendererHook.php:98]: I think this is superfluous.

File: library/Notifications/Hook/ObjectsRendererHook.php (hunk 5 above)
Comment (raviks789) [reply library/Notifications/Hook/ObjectsRendererHook.php:68]: After discussing it offline. I think, I understand this suggestion better.

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 8]
```diff
@@ -61,30 +64,58 @@ abstract public function getSourceType(): string;
...
+    private static function prepare(array $objectIdTags): void
     {
         $objectDisplayNames = [];
         foreach (Hook::all('Notifications\\ObjectsRenderer') as $hook) {
             /** @var self $hook */
             try {
-                $objectDisplayNames[] = $hook->getHtmlForObjectNames(self::$objectIdTagsCache[$hook->getSourceType()]);
+                $objectDisplayNames[] = $hook->getHtmlForObjectNames($objectIdTags[$hook->getSourceType()]);
+
+                unset($objectDisplayNames[$hook->getSourceType()]);
```
Comment (sukhwinder33445) [inline library/Notifications/Hook/ObjectsRendererHook.php:106]: I would prefer to reset the entire array after the loop. If there are 5 hooks, we call unset 5 times, which is unnecessary in my opinion.

**Review by sukhwinder33445 — DISMISSED** [review]

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+     *
+     * @return void
+     */
+    private static function prepare(array $objectIdTags): void
+    {
+        $objectDisplayNames = [];
+        foreach (Hook::all('Notifications\\ObjectsRenderer') as $hook) {
+            /** @var self $hook */
+            try {
+                $objectDisplayNames[] = $hook->getHtmlForObjectNames($objectIdTags[$hook->getSourceType()]);
```
Comment (sukhwinder33445) [inline library/Notifications/Hook/ObjectsRendererHook.php:106]: The call `$objectIdTags[$hook->getSourceType()]` must be fixed. It could throw `undefined array key` if there is no `icingadb` hook, for example.

File: library/Notifications/Widget/Detail/IncidentDetail.php [hunk 10]
```diff
@@ -52,39 +51,15 @@ protected function createContacts()
...
+        $objectUrl = ObjectsRendererHook::renderObjectLink($this->incident->object);
+
+        if (! $objectUrl) {
+            return [];
+        }
 
         return [
-            Html::tag('h2', t('Object')),
-            $list
+            new HtmlElement('h2', null, Text::create(t('Object'))),
```
Comment (sukhwinder33445) [inline library/Notifications/Widget/Detail/IncidentDetail.php:61]: Please use `Related Object` as content, just like in `EventDatail`.

**Review by sukhwinder33445 — COMMENTED** [review]
Please `squash` the commits.

**Review by sukhwinder33445 — APPROVED** [review]
LGTM

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+    protected static $objectNameHtmls = [];
+
+    /**
+     * Array of object names with their corresponding object IDs as keys
+     *
+     * It has the following structure : ['object ID' => 'object name'].
+     *
+     * @var array<string, string>
+     */
+    protected static $objectNames = [];
```
Comment (nilmerg) [inline library/Notifications/Hook/ObjectsRendererHook.php:49]: These should be private. I'd not want that implementations can change these.

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+
+    /**
+     * Get the object names for the objects using the object ID tags
+     *
+     * @param array<string, array<string, string>> $objectIdTags Array of object ID tags with their corresponding
```
Comment (nilmerg) [inline library/Notifications/Hook/ObjectsRendererHook.php:54]: the type doc doesn't match the description

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+     *                                                           ['object ID' => [object ID tags]]
+     *
+     * @return Generator<string, string> Generator for object names with their object ID tags as keys
+     */
+    abstract public function getObjectNames(array $objectIdTags): Generator;
+
+    /**
+     * Get the HTML for the object names for the objects using the object ID tags
+     *
+     * @param array<string, array<string, string>> $objectIdTags Array of object ID tags with their corresponding
```
Comment (nilmerg) [inline library/Notifications/Hook/ObjectsRendererHook.php:65]: the type doc doesn't match the description

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+
+        return self::createObjectNameAsString($obj);
+    }
+
+    private static function createObjectNameAsString(Objects $obj)
+    {
+        $objId = $obj->id;
+        if (isset(self::$objectNames[$objId])) {
+            return self::$objectNames[$objId];
+        }
```
Comment (nilmerg) [inline library/Notifications/Hook/ObjectsRendererHook.php:248]: Move this to `getObjectNameAsString`.

**Review by nilmerg — APPROVED** [review]

File: library/Notifications/Hook/ObjectsRendererHook.php [hunk 15]
```diff
@@ -277,13 +277,15 @@ final public static function renderObjectLink(Objects $object): ?ValidHtml
...
-                        ->addAttributes([
+                    $objectLink = $hook->createObjectLink($object->id_tags);
+                    if ($objectLink !== null) {
+                        return $objectLink->addAttributes([
                             'class' => [
                                 'icinga-module',
                                 'module-' . ($sourceType === 'icinga2' ? 'icingadb' : $sourceType)
                             ]
                         ]);
+                    }
```
Comment (nilmerg) [inline library/Notifications/Hook/ObjectsRendererHook.php:288]: The loop should end in this case, otherwise we allow multiple implementations for the same source type.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Notifications/Hook/ObjectsRendererHook.php (hunk 15 above)
Comment (sukhwinder33445) [reply library/Notifications/Hook/ObjectsRendererHook.php:288]: There is a return statement in the if condition.

Comment (nilmerg) [reply library/Notifications/Hook/ObjectsRendererHook.php:288]: That was previously the case. Now it's conditional.
